src/endpoint.rs

```diff
@@ -240,28 +339,46 @@ impl Endpoint {
             .connection_pool
             .insert(*node_addr, new_conn.connection.clone());
 
-        let conn = Connection::new(new_conn.connection, guard.clone());
-        let incoming_msgs = IncomingMessages::new(new_conn.uni_streams, new_conn.bi_streams, guard);
+        listen_for_incoming_messages(
+            new_conn.uni_streams,
+            new_conn.bi_streams,
+            guard,
+            self.message_tx.clone(),
+            self.disconnection_tx.clone(),
+        );
 
         self.connection_deduplicator
-            .complete(node_addr, Ok(conn.clone()))
+            .complete(node_addr, Ok(()))
             .await;
 
-        Ok((conn, Some(incoming_msgs)))
+        Ok(())
     }
 
-    /// Obtain stream of incoming QUIC connections
-    pub fn listen(&self) -> IncomingConnections {
-        trace!(
-            "Incoming connections will be received at {}",
-            self.local_addr()
-        );
-        IncomingConnections::new(self.quic_incoming.clone(), self.connection_pool.clone())
+    /// Get an existing connection for the peer address.
+    fn get_connection(&self, peer_addr: &SocketAddr) -> Option<Connection> {
+        if let Some((conn, guard)) = self.connection_pool.get(peer_addr) {
+            trace!("Using cached connection to peer: {}", peer_addr);
+            Some(Connection::new(conn, guard))
+        } else {
+            None
+        }
+    }
+
+    /// Sends a message to a peer. This will attempt to re-use any existing connections
+    /// with the said peer. If a connection doesn't exist already, a new connection will be created.
+    pub async fn send_message(&self, msg: Bytes, dest: &SocketAddr) -> Result<()> {
+        self.connect_to(dest).await?;
```

Comment:
Why not return the connection directly from `connect_to`? Then we wouldn't have to look it up again using `get_connection` here.

Reply:
The connection object is not a part of the public API. It is held internally and handled within this crate.